x-pack/legacy/plugins/ml/public/application/explorer/actions/load_explorer_data.ts

```diff
@@ -0,0 +1,327 @@
+
+import { i18n } from '@kbn/i18n';
+
+import { formatHumanReadableDateTime } from '../../util/date_utils';
+
+import {
+  explorerChartsContainerServiceFactory,
+  getDefaultChartsData,
+} from '../explorer_charts/explorer_charts_container_service';
+import { EXPLORER_ACTION, SWIMLANE_TYPE, VIEW_BY_JOB_LABEL } from '../explorer_constants';
+import { explorerAction$ } from '../explorer_dashboard_service';
+import {
+  getClearedSelectedAnomaliesState,
+  getDateFormatTz,
+  getDefaultSwimlaneData,
+  getSelectionInfluencers,
+  getSelectionTimeRange,
+  getSwimlaneBucketInterval,
+  getViewBySwimlaneOptions,
+  loadAnnotationsTableData,
+  loadAnomaliesTableData,
+  loadDataForCharts,
+  loadFilteredTopInfluencers,
+  loadOverallData,
+  loadTopInfluencers,
+  loadViewBySwimlane,
+  loadViewByTopFieldValuesForSelectedTime,
+} from '../explorer_utils';
+import { ExplorerState } from '../reducers';
+
+interface SwimlanePoint {
+  laneLabel: string;
+  time: number;
+}
+
+const memoizeIsEqual = (newArgs: any[], lastArgs: any[]) => isEqual(newArgs, lastArgs);
+
+// Memoize the data fetching methods
+// TODO: We need to track an attribute that allows refetching when the date picker
+// triggers a refresh, otherwise we'll get back the stale data. Note this was also
+// an issue with the previous version and the custom caching done within the component.
+const memoizedLoadAnnotationsTableData = memoizeOne(loadAnnotationsTableData, memoizeIsEqual);
+const memoizedLoadDataForCharts = memoizeOne(loadDataForCharts, memoizeIsEqual);
+const memoizedLoadFilteredTopInfluencers = memoizeOne(loadFilteredTopInfluencers, memoizeIsEqual);
+const memoizedLoadOverallData = memoizeOne(loadOverallData, memoizeIsEqual);
+const memoizedLoadTopInfluencers = memoizeOne(loadTopInfluencers, memoizeIsEqual);
+const memoizedLoadViewBySwimlane = memoizeOne(loadViewBySwimlane, memoizeIsEqual);
+const memoizedLoadAnomaliesTableData = memoizeOne(loadAnomaliesTableData, memoizeIsEqual);
+
+const dateFormatTz = getDateFormatTz();
+
+/**
+ * Fetches the data necessary for the Anomaly Explorer using observables.
+ *
+ * @param state ExplorerState
+ *
+ * @return action observable
+ */
+export function loadExplorerDataActionCreator(state: ExplorerState) {
+  const {
+    bounds,
+    filterActive,
+    filteredFields,
+    influencersFilterQuery,
+    isAndOperator,
+    noInfluencersConfigured,
+    selectedCells,
+    selectedJobs,
+    swimlaneContainerWidth,
+    swimlaneLimit,
+    tableInterval,
+    tableSeverity,
+    viewBySwimlaneFieldName: currentViewBySwimlaneFieldName,
+  } = state;
+
+  if (selectedJobs === null) {
+    return null;
+  }
+
+  const swimlaneBucketInterval = getSwimlaneBucketInterval(selectedJobs, swimlaneContainerWidth);
+
+  // TODO This factory should be refactored so we can load the charts using memoization.
+  const updateCharts = explorerChartsContainerServiceFactory(data => {
+    explorerAction$.next({
```

Comment:
you should probably have some `dispatch` method instead of directly update the Subject

Reply:
Good suggestion! In 6bad98d `explorer_dashboard_service.ts` now exposes a service with distinct methods to trigger specific actions. So we don't need to expose the `Subject` itself and the action constant anymore.